src/ELF/Builder.tcc

```diff
@@ -1097,6 +1117,146 @@ void Builder::build_dynamic_symbols(void) {
+  using Elf_Addr   = typename ELF_T::Elf_Addr;
+  using Elf_Xword  = typename ELF_T::Elf_Xword;
+  using Elf_Sxword = typename ELF_T::Elf_Sxword;
+
+  using Elf_Rela   = typename ELF_T::Elf_Rela;
+  using Elf_Rel    = typename ELF_T::Elf_Rel;
+  VLOG(VDEBUG) << "[+] Building object relocations";
+
+  it_object_relocations  object_relocations = this->binary_->object_relocations();
+
+  bool isRela = object_relocations[0].is_rela();
+  if (not std::all_of(
+        std::begin(object_relocations),
+        std::end(object_relocations),
+        [isRela] (const Relocation& relocation) {
+          return relocation.is_rela() == isRela;
+        })) {
+      throw LIEF::type_error("Object relocations are not of the same type");
+  }
+
+  it_sections sections = this->binary_->sections();
+
+  std::vector<Section*> rel_section;
+  for(Section& S: sections)
+    if(S.type() == ((isRela)?ELF_SECTION_TYPES::SHT_RELA:ELF_SECTION_TYPES::SHT_REL))
+      rel_section.push_back(&S);
+
+
+//  FIXME: Warn if not rel section found?
+
+  for(Section* section: rel_section) {
+
+    if (section->information() == 0 or section->information() >= sections.size())
+      throw LIEF::not_found("Unable to find associated section for SHT_REL{A} section");
+
+    const size_t sh_info = section->information();
+
+    Section& AssociatedSection = sections[sh_info];
+
+    std::vector<uint8_t> content;
+    for (const Relocation& relocation : this->binary_->object_relocations()) {
+
+      // Only write relocation in the matching section
+      // (relocation for .text in .rela.text)
+      // FIXME: static relocation on a new section will be ignored (SILENTLY!!)
+      if(relocation.section_ != &AssociatedSection)
+        continue;
+
+      // look for symbol index
+      uint32_t idx = 0;
+      if (relocation.has_symbol()) {
+        const Symbol& symbol    = relocation.symbol();
+        const std::string& name = symbol.name();
+        auto it_name  = std::find_if(
+            std::begin(this->binary_->dynamic_symbols_),
+            std::end(this->binary_->dynamic_symbols_),
+            [&name] (const Symbol* s) {
+            return s->name() == name;
+            });
+
+        if (it_name == std::end(this->binary_->dynamic_symbols_)) {
+          // FIXME: Do we have a way to walk both?
+          auto it_name  = std::find_if(
+              std::begin(this->binary_->static_symbols_),
+              std::end(this->binary_->static_symbols_),
+              [&name] (const Symbol* s) {
+              return s->name() == name;
+              });
+
+          if (it_name == std::end(this->binary_->static_symbols_)) {
+            throw not_found("Unable to find the symbol associated with the relocation");
+          }
+          idx = static_cast<uint32_t>(std::distance(std::begin(this->binary_->static_symbols_), it_name));
+        } else
+          idx = static_cast<uint32_t>(std::distance(std::begin(this->binary_->dynamic_symbols_), it_name));
```

Comment:
There is still an issue here as the idx may come from not a symbole but a section.

Reply:
Fixed